### ISSUE TITLE
Blocking equivalent to cancel effect?

### ISSUE THREAD

**Issue by Slapbox** [redux-saga/redux-saga#2312]
It would be really useful in our case to wait until the task is cancelled before continuing without needing a lot of code to make it blocking. Is there any interest in/plan for a blocking alternative to cancel?

The specific reason for us needing this is that we are cancelling a task that's has a pending `take` effect for the same value we're about to set via `put` in the next line.

**Comment by Andarist** [comment]
Isn't `join` already blocking and satisfying your needs?

Could you back up your idea with some code? It would be easier to grasp your situation and your requirements this way.

**Comment by Slapbox** [comment]
Thanks for your reply @Andarist. The issue with `join` is that, as I conceptualize of it, it can't handle the issue presented by the code below.

The core issue, is that the running task assigned to `handlerTask` must be cancellable from subsequent calls to `initiatorSaga`, or from alternative locations. I don't think `join` can handle this use case, but I'm probably mistaken.

For example:

1. `initiatorSaga` runs and forks `handler` task and assigns to `handlerTask`
2. `initiatorSaga` is run again and potentially needs to cancel the `handlerTask` from the previous run in a blocking fashion
3. OR `revertChanges` runs, which also needs to cancel `handlerTask` in a blocking fashion

Note: `try/catch` blocks excluded

```js
let handlerTask; // Defined outside saga function scope because it must be cancellable from subsequent calls to `initiatorSaga`, or from alternative locations.

function* handler() {
  yield put(actions.setValue({ foo: false }));
  yield take(
    action =>
      action.type === actions.setValue.type &&
      action.payload.foo
  ); // Wait for foo to be set anew
  // ... more logic
}

// `initiatorSaga` is potentially called multiple times, and needs to cancel `handler` that was forked from a different instance of `initiatorSaga`
function* initiatorSaga(val) {
  if (someCondition === true) {
    if (handlerTask) yield cancel(handlerTask); // Must run handler anew, and not have more than one of these tasks running at a time
    // TODO: Must wait for cancellation to occur before continuing
    handlerTask = yield fork(handler, val);
  } else {
    yield call(revertChanges);
  }
}

function* revertChanges(val) {
  yield cancel(handlerTask);
  // TODO: Must wait for cancellation to occur before continuing, because handler is waiting for foo to be set.
  yield put(
    actions.setValue({
      foo: val.foo,
    })
  );
}
````

**Comment by Slapbox** [comment]
Sorry to bump a topic so soon after posting, but it's the only remaining issue for us before we can release the next version of our app.

Does anyone have any thoughts on an approach for this that blocks execution until the task is cancelled? Or am I maybe/hopefully overlooking something?

**Comment by Slapbox** [comment]
Thank you for your suggestion @neurosnap; I'm so appreciative for your reply.

What you suggested is basically what we ended up doing, but it feels quite hacky indeed. I feel fortunate I only need this in two locations in our entire project or else the workaround could get out of hand.

Does the current design of `redux-saga` preclude the possibility of eventually implementing something like a blocking `cancel` effect available out of the box? I think it would be a great addition, but I don't know if it's plausible with the way the sagas work behind the scenes.

**Comment by neurosnap** [comment]
When you cancel an effect it potentially has to cancel a bunch of effects (up the tree).  A quick glance at the code seems like it should theoretically be possible to have a blocking `cancel` effect but we'd have to ensure that all calls to `.cancel` on the effect are resolved before returning.  

Because of the web of how `cancel` gets called up the tree, it might result in a breaking change where we have to convert `cancel` from non-blocking to blocking.  That kind of breaking change would be tough for us to justify at this stage in the life of the project.  However, I haven't delved too deeply into this part of the codebase so take what I'm saying with a grain of salt.

**Comment by neurosnap** [comment]
It looks like we've come to a conclusion for this issue.  Feel free to open a separate issue to discuss any other issues!
